src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs

```diff
@@ -502,42 +492,52 @@ internal static void CreateSymbolicLink(string path, string pathToTarget, bool i
-                // Another nuance is that SubstituteName does not contain redundant path segments while PrintName does.
-                // PrintName can ONLY return a NT path if the link was created explicitly targeting a file/folder in such way.
-                //   e.g: mklink /D linkName \??\C:\path\to\target.
+                // We always use SubstituteName(Offset|Length) instead of PrintName(Offset|Length),
+                // the latter is just the display name of the reparse point and it can show something completely unrelated to the target.
 
                 if (rbSymlink.ReparseTag == Interop.Kernel32.IOReparseOptions.IO_REPARSE_TAG_SYMLINK)
                 {
-                    int printNameOffset = sizeof(Interop.Kernel32.SymbolicLinkReparseBuffer) + rbSymlink.PrintNameOffset;
-                    int printNameLength = rbSymlink.PrintNameLength;
+                    int offset = sizeof(Interop.Kernel32.SymbolicLinkReparseBuffer) + rbSymlink.SubstituteNameOffset;
+                    int length = rbSymlink.SubstituteNameLength;
 
-                    Span<char> targetPath = MemoryMarshal.Cast<byte, char>(bufferSpan.Slice(printNameOffset, printNameLength));
-                    Debug.Assert((rbSymlink.Flags & Interop.Kernel32.SYMLINK_FLAG_RELATIVE) == 0 || !PathInternal.IsExtended(targetPath));
+                    Span<char> targetPath = MemoryMarshal.Cast<byte, char>(bufferSpan.Slice(offset, length));
 
-                    if (returnFullPath && (rbSymlink.Flags & Interop.Kernel32.SYMLINK_FLAG_RELATIVE) != 0)
+                    bool isRelative = (rbSymlink.Flags & Interop.Kernel32.SYMLINK_FLAG_RELATIVE) != 0;
+                    if (!isRelative)
                     {
-                        // Target path is relative and is for ResolveLinkTarget(), we need to append the link directory.
-                        return Path.Join(Path.GetDirectoryName(linkPath.AsSpan()), targetPath);
-                    }
+                        // Absolute target is in NT format and we need to clean it up before return it to the user.
+                        if (targetPath.StartsWith(PathInternal.UncNTPathPrefix.AsSpan()))
+                        {
+                            // We need to prepend the Win32 equivalent of UNC NT prefix.
+                            return Path.Join(PathInternal.UncPathPrefix.AsSpan(), targetPath.Slice(PathInternal.UncNTPathPrefix.Length)).ToString();
+                        }
 
-                    return targetPath.ToString();
+                        Debug.Assert(targetPath.StartsWith(PathInternal.NTPathPrefix.AsSpan()));
+                        return targetPath.Slice(PathInternal.NTPathPrefix.Length).ToString();
+                    }
+                    else if (returnFullPath)
+                    {
+                        return Path.Join(Path.GetDirectoryName(linkPath.AsSpan()), targetPath).ToString();
```

Comment:
The `ToString()` is unnecessary. All the `Path.Join` overloads return a `string`.

Reply:
Maybe we should have an analyzer to catch calling ToString() on a string..

Reply:
> Maybe we should have an analyzer to catch calling ToString() on a string..

It has a side effect: it performs a null check. So it can't always be removed. Regardless, I see this in generated code sometimes, where the generator just blindly calls `ToString` on data regardless of the data's underlying type, and it's pretty harmless in those scenarios.